src/stream.rs

```diff
@@ -0,0 +1,229 @@
+// Copyright 2019, The Gtk-rs Project Developers.
+// See the COPYRIGHT file at the top-level directory of this distribution.
+// Licensed under the MIT license, see the LICENSE file or <http://opensource.org/licenses/MIT>
+
+use ::ffi::{self, cairo_status_t};
+use ::{Status, Surface, UserDataKey};
+
+use libc::{c_void, c_double, c_uchar, c_uint};
+use std::any::Any;
+use std::cell::UnsafeCell;
+use std::io;
+use std::panic::AssertUnwindSafe;
+use std::rc::Rc;
+
+macro_rules! for_stream_constructors {
+    ($constructor_ffi: ident) => {
+        /// Takes full ownership of the output stream,
+        /// which is not allowed to borrow any lifetime shorter than `'static`.
+        ///
+        /// Because the underlying `cairo_surface_t` is reference-counted,
+        /// a lifetime parameter in a Rust wrapper type would not be enough to track
+        /// how long it can keep writing to the stream.
+        pub fn for_stream<W: io::Write + 'static>(width: f64, height: f64, stream: W) -> Self {
+            Self {
+                inner: Surface::_for_stream(
+                    ffi::$constructor_ffi,
+                    width,
+                    height,
+                    stream,
+                ),
+            }
+        }
+
+        /// Allows writing to a borrowed stream. The lifetime of the borrow is not tracked.
+        ///
+        /// # Safety
+        ///
+        /// The value that `stream` points to must live at least until the underlying `cairo_surface_t`
+        /// (which maybe be longer then the Rust `PdfSurface` wrapper, because of reference-counting),
+        /// or until the output stream is removed from the surface with [`Surface::take_output_stream`].
+        ///
+        /// Since the former is hard to track for sure, the latter is strongly recommended.
+        /// The concrete type behind the `Box<dyn Any>` value returned by `take_output_stream`
+        /// is private, so you won’t be able to downcast it.
+        /// But removing it anyway ensures that later writes do no go through a dangling pointer.
+        pub unsafe fn for_raw_stream<W: io::Write + 'static>(width: f64, height: f64, stream: *mut W) -> Self {
```

Comment:
Should we also remove the `'static` here to make it more generic and unsafe? :)

Reply:
We can’t easily. That bound is required to keep even `*mut W` in a `Box<dyn Any>`.

Reply:
Makes sense, thanks